### ISSUE TITLE
[regression/8.0.0] TapGestureRecognizers not working in maui embedded

### ISSUE THREAD

**Issue by ultimategrandson** [dotnet/maui#17948]
### Description

TapGestureRecognizer used to work in Xamarin.Forms just fine.

I migrated to .NET 8.0 and MAUI, TapGestureRecognizers kind of worked. If I used them on a Frame control, it didn't fire, so I changed everything to use the Border control and that worked.

Now, I have migrated to .NET 8.0 and TapGestureRecognizers don't seem to work at all.

I created a test app, where I added a Border with a Grid inside and a TapGestureRecognizer, the Tapped event seems to work in this case, but the Command does not. 

In my main project, I am using Maui Embedded and it doesn't work with the Tapped event nor the Command.

```xaml
<Border BackgroundColor="{DynamicResource Accent}"
        MaximumWidthRequest="300"
        HeightRequest="40"
        Margin="20, 0, 20, 5"
        x:Name="LoginButton">

    <Border.StrokeShape>
        <RoundRectangle CornerRadius="4" />
    </Border.StrokeShape>

    <Border.GestureRecognizers>
        <TapGestureRecognizer Command="{Binding LoginButtonTappedCommand}" />
    </Border.GestureRecognizers>

    <Grid>
        <Grid.ColumnDefinitions>
            <ColumnDefinition Width="Auto" />
            <ColumnDefinition Width="*" />
            <ColumnDefinition Width="Auto" />
            <ColumnDefinition Width="Auto" />
        </Grid.ColumnDefinitions>

        <Image Grid.Column="0"
            Source="fingerprint.png"
            HorizontalOptions="Start"
            VerticalOptions="Center"
            Margin="5,0,0,0"
            IsVisible="false"
            x:Name="FingerprintIcon" />

        <Label Grid.Column="1"
            Text="Log In"
            TextColor="White"
            VerticalTextAlignment="Center"
            Margin="20,0"
            x:Name="LoginButtonText" />

        <ActivityIndicator Grid.Column="2" 
            x:Name="LoginActivityIndicator"
            HorizontalOptions="End"
            Scale="0.7"
            Color="White"
            IsVisible="{Binding IsLoading}"
            IsRunning="{Binding IsLoading}" />

        <Image Grid.Column="3"
            Source="chevron_right.png"
            VerticalOptions="Fill"
            HorizontalOptions="End"
            Margin="0, 0, 5, 0">
            <Image.Behaviors>
                <behaviors:IconTintColorBehavior TintColor="White" />
            </Image.Behaviors>
        </Image>
    </Grid>
</Border>
```

Is there some documentation on what has changed? Or is this just a bug?

### Version with bug

8.0.0-rc.1.9171

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI, Yes, this used to work in Xamarin.Forms

### Last version that worked well

7.0.92

### Affected platforms

iOS, Android

### Did you find any workaround?

https://github.com/dotnet/maui/issues/17948#issuecomment-1828833433

**Comment by Symbai** [comment]
> Or is this just a bug?

It's a bug, see #16622 and the linked mega discussion where some people reported some workarounds which works in some cases. Still it's not fixed and sometimes no workaround exist.

**Comment by ultimategrandson** [comment]
#8895 is closed and it sounds like a lot of the issues should have been resolved?

**Comment by Symbai** [comment]
Yes its closed, due to the fact that several people reported several different issues on several different controls. We were told to reopen if there is still a problem. As in #16622 you can see TapGestureRecognizer is still bugged (and confirmed its still bugged) which is also what you experience.

**Comment by mattleibow** [comment]
Is it possible to attach a repro? I have copied your XAML and tried to have a look but RC 2 seems to be working for me:

```cs
public partial class MainPage : ContentPage
{
	public MainPage()
	{
		InitializeComponent();
		BindingContext = this;
	}

	public ICommand LoginButtonTappedCommand { get; } = new Command(() =>
	{
		Console.WriteLine("CLICKED LoginButtonTappedCommand");
	});
}
```

```xaml
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="TestiPadCrash123.MainPage" Background="Azure">

    <ScrollView>
        <VerticalStackLayout
            Spacing="25"
            Padding="30,0"
            VerticalOptions="Center">

            <Border BackgroundColor="{DynamicResource Accent}"
                    MaximumWidthRequest="300"
                    HeightRequest="40"
                    Margin="20, 0, 20, 5"
                    x:Name="LoginButton">

                <Border.StrokeShape>
                    <RoundRectangle CornerRadius="4" />
                </Border.StrokeShape>

                <Border.GestureRecognizers>
                    <TapGestureRecognizer Command="{Binding LoginButtonTappedCommand}" />
                </Border.GestureRecognizers>

                <Grid>
                    <Grid.ColumnDefinitions>
                        <ColumnDefinition Width="Auto" />
                        <ColumnDefinition Width="*" />
                        <ColumnDefinition Width="Auto" />
                        <ColumnDefinition Width="Auto" />
                    </Grid.ColumnDefinitions>

                    <Image Grid.Column="0"
                        Source="dotnet_bot.png"
                        HorizontalOptions="Start"
                        VerticalOptions="Center"
                        Margin="5,0,0,0"
                        IsVisible="false"
                        x:Name="FingerprintIcon" />

                    <Label Grid.Column="1"
                        Text="Log In"
                        TextColor="White"
                        VerticalTextAlignment="Center"
                        Margin="20,0"
                        x:Name="LoginButtonText" />

                    <ActivityIndicator Grid.Column="2" 
                        x:Name="LoginActivityIndicator"
                        HorizontalOptions="End"
                        Scale="0.7"
                        Color="White"
                        IsVisible="{Binding IsLoading}"
                        IsRunning="{Binding IsLoading}" />

                    <Image Grid.Column="3"
                        Source="dotnet_bot.png"
                        VerticalOptions="Fill"
                        HorizontalOptions="End"
                        Margin="0, 0, 5, 0">
                    </Image>
                </Grid>
            </Border>

        </VerticalStackLayout>
    </ScrollView>

</ContentPage>
```

**Comment by anotherlab** [comment]
I'm having the same issue.  Code that worked in Xamarin.Forms 5 and .NET 7 does not in 8 RC 1 or 8 RC 2.

**Comment by ultimategrandson** [comment]
https://github.com/ultimategrandson/MauiEmbeddedTapGestureBrokenRepro

I tried to create the simplest repro, run it on Android and try and click the black label, and it doesn't work. 

For me, its because I use Maui Embedded. Our app has been around since before Xamarin Forms was a thing, some parts of our App then used Xamarin Forms because it made dev faster. Now we're struggling to have a stable version that uses Maui because not enough testing seems to be done in embedded scenarios. 

It would be great to use Maui to replace all our other platform specific pages/forms. However, we need to have the embedded features stable first so we can reliable maintain what we have now before we can migrate to a purely Maui app.

**Comment by anotherlab** [comment]
@mattleibow I could not get your example to trigger the TapGestureRecognizer.  Even if it did work, I need to use a TapGestureRecognizer in a list and pass in the item being tapped and my use case did not your example.

I created a repo at https://github.com/anotherlab/MauiTapGestureRecognizersNotWorking that shows code that will work with .NET 7, but not with .NET 8

**Comment by mattleibow** [comment]
@anotherlab I think you may be seeing another issue than OP.

When running your sample on mac, I saw:

> [0:] Microsoft.Maui.Controls.Xaml.Diagnostics.BindingDiagnostics: Warning: 'EmployeeTapped' property not found on 'TapGestureIssue.EmployeesViewModel', target property: 'Microsoft.Maui.Controls.TapGestureRecognizer.Command'

I then changed the property to be an instance property:

<img width="482" alt="image" src="https://github.com/dotnet/maui/assets/1096616/5daffd0a-ea8e-4ede-a0a2-07e6f16825c5">

After that, the sample app worked on both macOS and Android:


https://github.com/dotnet/maui/assets/1096616/214e50d3-cca0-4896-a717-a4049b790e28

I am using .NET 8 RC 2:

```
Installed Workload Id      Manifest Version                     Installation Source
-----------------------------------------------------------------------------------
macos                      13.3.8968-net8-rc2/8.0.100-rc.2      SDK 8.0.100-rc.2   
ios                        16.4.8968-net8-rc2/8.0.100-rc.2      SDK 8.0.100-rc.2   
maccatalyst                16.4.8968-net8-rc2/8.0.100-rc.2      SDK 8.0.100-rc.2   
maui                       8.0.0-rc.2.9373/8.0.100-rc.2         SDK 8.0.100-rc.2   
tvos                       16.4.8968-net8-rc2/8.0.100-rc.2      SDK 8.0.100-rc.2   
android                    34.0.0-rc.2.468/8.0.100-rc.2         SDK 8.0.100-rc.2  
```

Since I believe your issue is different to @ultimategrandson and I can reproduce the embedded failure, please open a new issue so that one of the issues is not lost accidentially.

**Comment by mattleibow** [comment]
@ultimategrandson thanks for the repo, I am able to reproduce your issue. @PureWeen not sure how this relates to embedding, but might be the contexts not set up right.

Do we have an embedded sample? I know you were working on things, so checking to see if you had one? If not, we need to create one and make sure it gets added to the maui repo. I can do the last part.

**Comment by Symbai** [comment]
If someone checks it out, before creating a PR please also take a look at #16622 (containing a repro). I really don't want to "hijack" this issue here but also don't want mine getting closed again as "We recently fixed some TapGestures issues, please reopen if it still exist".

**Comment by jsuarezruiz** [comment]
I've been reviewing the example. The NET MAUI's internal GestureManager is processing gestures correctly. The problem is that when try to invoke the command it does nothing as it is null.

There are options where this is expected, for example:
`Command="{Binding EmployeeTapped}"`

The BndingContext of the gesture is Employee where the command does not exist. But in other cases, where use the context of the CollectionView or similar, it should work. The command is static, is it for some particular reason? Removing static from the command, all works as expected.

![image](https://github.com/dotnet/maui/assets/6755973/0580d77d-a64e-4392-b523-9c3a457cc170)

**Comment by samhouts** [comment]
@StephaneDelcroix When we're looking up the methods on a DataContext for binding to commands, are we now excluding the static methods?

**Comment by samhouts** [comment]
Okay, we got distracted by the static methods part, so I moved that to a new issue and hid the comments on this issue so we can stay focused on @ultimategrandson's report. Thanks!

**Comment by artemvalieiev** [comment]
We currently migrating our app from native views to MAUI and want to use CollectionView's and other improvements that are available only .NET8, but this issue is stopping us from upgrading to .net8. Is there any workaround for this issue so we can apply it on the client side and don't need to wait until SR1 release?

**Comment by artemvalieiev** [comment]
I tried to investigate this problem. GestureManager is always disposed of automatically after creation because the Window property of the elements is always null on iOS https://github.com/dotnet/maui/blob/b41492592d717e49883b667e8b254b8ba3f4de8d/src/Controls/src/Core/Platform/GestureManager/GestureManager.cs#L55-L65

If I manually create the window with the page I want and set it's WindowHandler, gestures work again. While this makes the TapGestureGalleryPage sample page work, I don't know if it is a valid workaround we can use in production with upcoming .NET8 until SR1 will fix this

```
var mauiWindow = new Window { Page = new TapGestureGalleryPage()};
var yourAppDelegate = (AppDelegate)UIApplication.SharedApplication.Delegate;
var mauiContext = yourAppDelegate.MauiContext!;
yourAppDelegate.Window!.SetWindowHandler(mauiWindow, mauiContext);

var page = mauiWindow.Page!;
UIViewController myPageController = page.ToUIViewController(mauiContext);
```

**Comment by artemvalieiev** [comment]
@fedemkr I haven't tried the app extension yet, but for the regular .net8-ios app:

1. You still need to register mauiAppBuilder.Services.Add(new ServiceDescriptor(typeof(UIWindow), UIApplication.SharedApplication.KeyWindow)); as you said
2. Register of `CustomWindowHandler` provided in MauiProgram via `ConfigureMauiHandlers` when you create MauiApp via builder

```
/// <summary>
/// Haven't fully tested this yet
/// </summary>
public class CustomWindowHandler : ElementHandler<IWindow, UIWindow>, IWindowHandler
{
    public static IPropertyMapper<IWindow, IWindowHandler> Mapper = new PropertyMapper<IWindow, IWindowHandler>(ElementHandler.ElementMapper)
    {
    };
    
    public CustomWindowHandler() : base(Mapper)
    {
    }
    
    protected override UIWindow CreatePlatformElement()
    {
	    // Haven't tested
	    return UIApplication.SharedApplication.Delegate.GetWindow();
    }
}
```
3. Override OnHandlerChanged of the page you want to display with 
```
protected override void OnHandlerChanged()
{
  base.OnHandlerChanged();
  
  var window = new Window();
  window.ToHandler(this.Handler.MauiContext);
  window.Page = this;
}
```
We use only a ContentPage, not NavigationPage. But since you use NavigationPage as a ToUIViewController target, you don't need a window fix in pages. Create a subclass of NavigationPage, add Window set fix and other pages that will be used inside that NavigationPage, automatically receive window property, and touch gestures will start to work again.
```
public class NavigationPageFix : NavigationPage
{
    public NavigationPageFix()
    {
    }

    public NavigationPageFix(Page root) : base(root)
    {
    }

    protected override void OnHandlerChanged()
    {
        base.OnHandlerChanged();

        var window = new Window();
        window.ToHandler(this.Handler.MauiContext);
        window.Page = this;
    }
}
```

**Comment by fedemkr** [comment]
Hi there, thanks for the help but sadly I can't make it work.
I have tried these combinations:

**Solution proposed by @artemvalieiev with the NavigationPageFix**

I'm getting the next error:

```
System.InvalidOperationException: No page was set on the window.
   at Microsoft.Maui.Controls.Window.Microsoft.Maui.IWindow.get_Content()
   at Microsoft.Maui.Handlers.WindowHandler.MapContent(IWindowHandler handler, IWindow window)
   at Microsoft.Maui.PropertyMapper`2.<>c__DisplayClass5_0[[Microsoft.Maui.IWindow, Microsoft.Maui, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null],[Microsoft.Maui.Handlers.IWindowHandler, Microsoft.Maui, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].<Add>b__0(IElementHandler h, IElement v)
   at Microsoft.Maui.PropertyMapper.UpdatePropertyCore(String key, IElementHandler viewHandler, IElement virtualView)
   at Microsoft.Maui.PropertyMapper.UpdateProperties(IElementHandler viewHandler, IElement virtualView)
   at Microsoft.Maui.Handlers.ElementHandler.SetVirtualView(IElement view)
   at Microsoft.Maui.Controls.Element.SetHandler(IElementHandler newHandler)
   at Microsoft.Maui.Controls.Element.set_Handler(IElementHandler value)
   at Microsoft.Maui.Platform.ElementExtensions.ToHandler(IElement view, IMauiContext context)
   at Com.iOS.SomeExtension.NavigationPageFix.OnHandlerChanged() in /somePath.../MyProject/src/iOS.SomeExtension/CustomViewController.cs:line 41
   at Microsoft.Maui.Controls.Element.OnHandlerChangedCore()
   at Microsoft.Maui.Controls.VisualElement.OnHandlerChangedCore()
   at Microsoft.Maui.Controls.NavigationPage.OnHandlerChangedCore()
   at Microsoft.Maui.Controls.Element.SetHandler(IElementHandler newHandler)
   at Microsoft.Maui.Controls.Element.set_Handler(IElementHandler value)
   at Microsoft.Maui.Controls.VisualElement.Microsoft.Maui.IElement.set_Handler(IElementHandler value)
   at Microsoft.Maui.Platform.ElementExtensions.ToHandler(IElement view, IMauiContext context)
   at Microsoft.Maui.Platform.ElementExtensions.ToPlatform(IElement view, IMauiContext context)
   at Microsoft.Maui.Platform.ElementExtensions.ToUIViewController(IElement view, IMauiContext context)
   at Com.iOS.SomeExtension.CustomViewController.LaunchHomePage() in /somePath.../MyProject/src/iOS.SomeExtension/CustomViewController.cs:line 454
```

**Solution proposed by @artemvalieiev with the NavigationPageFix but inverting the order of setting the page**

```c#
    protected override void OnHandlerChanged()
    {
        base.OnHandlerChanged();

        var window = new Window();
        window.Page = this;
        window.ToHandler(this.Handler.MauiContext);
    }
```
or
```c#
    protected override void OnHandlerChanged()
    {
        base.OnHandlerChanged();

        var window = new Window(this);
        window.ToHandler(this.Handler.MauiContext);
    }
```

I get the following error:
```
System.InvalidOperationException: MauiContext is null.
   at Microsoft.Maui.Controls.Window.get_MauiContext()
   at Microsoft.Maui.Controls.Platform.AlertManager.Subscribe(Window window)
   at Microsoft.Maui.Controls.Platform.AlertManager.Subscribe()
   at Microsoft.Maui.Controls.Window.OnPageHandlerChanged(Object sender, EventArgs e)
   at Microsoft.Maui.Controls.Window.OnPageChanged(Page oldPage, Page newPage)
   at Microsoft.Maui.Controls.Window.<>c.<.cctor>b__211_0(BindableObject b, Object o, Object n)
   at Microsoft.Maui.Controls.BindableObject.SetValueActual(BindableProperty property, BindablePropertyContext context, Object value, Boolean currentlyApplying, SetValueFlags attributes, SetterSpecificity specificity, Boolean silent)
   at Microsoft.Maui.Controls.BindableObject.SetValueCore(BindableProperty property, Object value, SetValueFlags attributes, SetValuePrivateFlags privateAttributes, SetterSpecificity specificity)
   at Microsoft.Maui.Controls.BindableObject.SetValue(BindableProperty property, Object value)
   at Microsoft.Maui.Controls.Window.set_Page(Page value)
   at Com.iOS.SomeExtension.NavigationPageFix.OnHandlerChanged() in /somePath.../MyProject/src/iOS.SomeExtension/CustomViewController..cs:line 41
   at Microsoft.Maui.Controls.Element.OnHandlerChangedCore()
   at Microsoft.Maui.Controls.VisualElement.OnHandlerChangedCore()
   at Microsoft.Maui.Controls.NavigationPage.OnHandlerChangedCore()
   at Microsoft.Maui.Controls.Element.SetHandler(IElementHandler newHandler)
   at Microsoft.Maui.Controls.Element.set_Handler(IElementHandler value)
   at Microsoft.Maui.Controls.VisualElement.Microsoft.Maui.IElement.set_Handler(IElementHandler value)
   at Microsoft.Maui.Platform.ElementExtensions.ToHandler(IElement view, IMauiContext context)
   at Microsoft.Maui.Platform.ElementExtensions.ToPlatform(IElement view, IMauiContext context)
   at Microsoft.Maui.Platform.ElementExtensions.ToUIViewController(IElement view, IMauiContext context)
   at Com.iOS.SomeExtension.CustomViewController.LaunchHomePage() in /somePath.../MyProject/src/iOS.SomeExtension/CustomViewController..cs:line 454
```

**Same solutions but changing the MauiContext**

If I change
`this.Handler.MauiContext`
for
`MauiContextSingleton.Instance.MauiContext` (which is the one I create on the extension)

Then the same aforementioned errors happen.

**Same solutions but changing CustomWindowHandler**

If I change
```c#
    protected override UIWindow CreatePlatformElement()
    {
	    // Haven't tested
	    return UIApplication.SharedApplication.Delegate.GetWindow();
    }
```
for
```c#
    protected override UIWindow CreatePlatformElement()
    {
	    // Haven't tested
            return Platform.GetCurrentUIViewController().View.Window;
    }
```

Then the same aforementioned errors happen.

----
Just in case I'm creating the maui context as follows:
```c#
var builder = MauiApp.CreateBuilder()
                          // some other stuff
                          .UseMauiEmbedding<Application>();

// ...

builder.Services.Add(new ServiceDescriptor(typeof(UIWindow), UIApplication.SharedApplication.KeyWindow));
var mauiApp = builder.Build();
MauiContextSingleton.Instance.Init(new MauiContext(mauiApp.Services));
```

-----

Am I missing setting the `MauiContext` elsewhere? Is there another workaround I can try?

Thanks for everything!

**Comment by PureWeen** [comment]
@fedemkr I updated the workaround here to use a NavigationPage

https://github.com/dotnet/maui/issues/17948#issuecomment-1828833433

If you have a repro you can package up, I can see what's not quite working in your scenario.

**Comment by ultimategrandson** [comment]
> > https://github.com/ultimategrandson/MauiEmbeddedTapGestureBrokenRepro
> > I tried to create the simplest repro, run it on Android and try and click the black label, and it doesn't work.
> > For me, its because I use Maui Embedded. Our app has been around since before Xamarin Forms was a thing, some parts of our App then used Xamarin Forms because it made dev faster. Now we're struggling to have a stable version that uses Maui because not enough testing seems to be done in embedded scenarios.
> > It would be great to use Maui to replace all our other platform specific pages/forms. However, we need to have the embedded features stable first so we can reliable maintain what we have now before we can migrate to a purely Maui app.
> 
> Can you try the workarounds I have here?
> 
> [Archive.zip](https://github.com/dotnet/maui/files/13507073/Archive.zip)

This seems to be working!

With iOS, I added the code into ViewDidLoad to prepare the page. However, if I wrap the page within the NavigationPage, I get a 
 "UIViewControllerHierarchyInconsistency Reason: A view can only be associated with at most one view controller at a time!" error when pushing a new controller with a different Maui ContentPage. I didn't understand why this error is occurring since they are two separate controllers. Removing the NavigationPage stopped the error occurring.

What's the reasoning behind using the NavigationPage?

**Comment by PureWeen** [comment]
@ultimategrandson I just added a navpage because @fedemkr was asking

You don't have to add one

**Comment by muak** [comment]
I am publishing a package called [SettingsView](https://github.com/muak/AiForms.Maui.SettingsView/tree/dev/net8), which is not using Maui emmbedded, but internally uses a ContentView as a native View by using ToHandler().
I am having the same problem there, GestureRecognizer is not working at all.

I can make it work by setting the ParentPage to the corresponding View's Parent, referring to the workaround here, but that would ignore the original tree structure.

If the following,

- Page
  - ViewA
    - ViewB  (targetView)

the gesture is not working.

if the following works fine.

- Page
  - ViewB (targetView)

This does not occur in .NET7, but only in .NET8.
